code/pngutils/pngutils.cpp

```diff
@@ -206,3 +210,547 @@ int png_read_bitmap(const char *real_filename, ubyte *image_data, ubyte *bpp, in
+	char *p = strchr( filename, '.' );
+	if ( p != nullptr ) *p = 0;
+	strcat_s( filename, ".png" );
+
+	_cfp = cfopen( filename , "rb" );
+
+	if ( _cfp == nullptr) {
+		_apng_failed("couldn't open filename");
+	}
+
+	_reading = true;
+
+	ubyte sig[8];
+	if (cfread(sig, 8, 1, _cfp) != 1)  {
+		_apng_failed("cfread of png signature failed");
+	}
+	if (png_sig_cmp(sig, 0, 8) != 0) {
+		_apng_failed("file has invalid png signature");
+	}
+
+	_id = _read_chunk(_chunk_IHDR);
+
+	if (_id != id_IHDR || _chunk_IHDR.size != 25) {
+		_apng_failed("failed to read IHDR chunk");
+	}
+
+	w = png_get_uint_32(&_chunk_IHDR.data[8]);
+	h = png_get_uint_32(&_chunk_IHDR.data[12]);
+	_row_len = w * 4;
+	bpp = 32;  // we'll force all our frames to use this
+
+	// setup frames & keep bm_create happy
+	_image_size = _row_len * h;
+	frame.data.reserve(_image_size); // alloc only once
+	frame.data.assign(_image_size, 0); // all transparent black per spec
+	frame.rows.resize(h);
+	_frame_raw.data.resize(_image_size);
+	_frame_raw.rows.resize(h);
+	_frame_next.data.resize(_image_size);
+	_frame_next.rows.resize(h);
+	for (uint i = 0; i < h; ++i) {
+		frame.rows.at(i) =       &frame.data.at(i * _row_len);
+		_frame_raw.rows.at(i) =  &_frame_raw.data.at(i * _row_len);
+		_frame_next.rows.at(i) = &_frame_next.data.at(i * _row_len);
+	}
+
+	// read all data
+	while (!cfeof(_cfp)) {
+		_process_chunk();
+	}
+
+	// should be at EOF; attach to _frame_offsets to make next_frame code simpler
+	Assertion(cfeof(_cfp) != 0, "apng not at EOF, get a coder!");
+	_frame_offsets.push_back(cftell(_cfp));
+
+	// sanity checks
+	if (anim_time <= 0.0f) {
+		_apng_failed("animation duration <= 0.0f, bad data?");
+	}
+
+	if (nframes < 1) {
+		_apng_failed("animation didn't have any frames, is this a static png?");
+	}
+
+	// reset _cfp so it can be used for next frame
+	_reading = false;
+	if (cfseek(_cfp, _frame_offsets.at(0), CF_SEEK_SET) != 0) {
+		_apng_failed("couldn't seek to 1st fcTL offset");
+	}
+
+	return PNG_ERROR_NONE;
+}
+
+/*
+ * @brief something went badly wrong, throw an exception
+ *
+ * @param [in] msg  text to display about the error
+ */
+void apng_ani::_apng_failed(const char* msg)
+{
+	nframes = 0;
+	current_frame = 0;
+	plays = 0;
+	anim_time = 0.0f;
+
+	SCP_string error_msg = "(file ";
+	error_msg += _filename;
+	error_msg += ") ";
+	error_msg += msg;
+	throw ApngException(error_msg.c_str());
+}
+
+/*
+ * @brief ctor
+ *
+ * @note loads apng header data
+ */
+apng_ani::apng_ani(const char* filename)
+	: w(0)
+	, h(0)
+	, bpp(0)
+	, nframes(0)
+	, current_frame(0)
+	, plays(0)
+	, anim_time(0.0f)
+	, _filename(filename)
+	, _pngp(nullptr)
+	, _infop(nullptr)
+	, _cfp(nullptr)
+	, _offset(0)
+	, _sequence_num(0)
+	, _id(0)
+	, _row_len(0)
+	, _image_size(0)
+	, _framew(0)
+	, _frameh(0)
+	, _x_offset(0)
+	, _y_offset(0)
+	, _delay_num(1)
+	, _delay_den(100)
+	, _dispose_op(0)
+	, _blend_op(0)
+	, _reading(true)
+	, _got_acTL(false)
+	, _got_IDAT(false)
+{
+	load_header();
```

Comment:
Just checking, you mean call load_header outside of the constructor?

I thought that when exceptions where thrown the destructor was always called, but I guess that due to the issue Axem ran into I must be wrong.

Anyway, slightly irritating because I was relying on passing the filename in to fully prep the apng for use, but oh well, it's just one extra line of code each time the class is constructed. Although... would destroying the png/cfile objects within the exception handler (_apng_failed) also work? (although that's another duplication of code)